src/primitive_file_non_win.cc

```diff
@@ -438,6 +442,47 @@ PRIMITIVE(chdir) {
 #endif
 }
 
+PRIMITIVE(chmod) {
+  ARGS(cstring, pathname, int, mode);
+  int result = FILE_CHMOD_(current_dir(process), pathname, mode, 0);
+  if (result < 0) return return_open_error(process, errno);
+  return process->null_object();
+}
+
+// Notice that link does not use the *at(fd,...) version, as the symlink call does not
+// support it. Therefore, source and target must be absolute, this should be handled in the
+// toit api code.
+PRIMITIVE(link) {
+  ARGS(cstring, source, cstring, target, int, type);
+  int result;
+  if (type == 0) {
+    result = link(target, source);
+  } else { // type 1 and 2 are only different on windows.
+    result = symlink(target, source);
+  }
+  if (result < 0) return return_open_error(process, errno);
+  return process->null_object();
+}
+
+PRIMITIVE(readlink) {
+  ARGS(cstring, pathname);
+
+  AllocationManager allocation(process);
+  uint8* backing = allocation.alloc(PATH_MAX);
+  if (!backing) FAIL(ALLOCATION_FAILED);
+
+  int result = FILE_READLINK_(process->current_directory(), pathname,
+                              reinterpret_cast<char*>(backing), PATH_MAX);
+  if (result < 0) return return_open_error(process, errno);
+
+  String* string = process->allocate_string(result);
+  if (!string) FAIL(ALLOCATION_FAILED);
+
+  strncpy(string->as_cstr(), reinterpret_cast<char*>(backing), result);
```

Comment:
Please use a String::MutableBytes object here.  The as_cstr is intended for getting a null terminated char* out of a string for passing to C APIs.